### PR TITLE
perf(sequence-packing): use CPU sequence boundaries

### PR DESCRIPTION
**Pull request by CarlosGomes98**
# What does this PR do ?

- Convert many separate CPU-GPU syncs to 1-time bulk syncs
- Move all required GPU tensors to host before entering the loss and logprob code. This eliminates repeated D2H copies
- Introduce the to_cpu_int_tuple helper and CpuIntTuple type to establish a clear contract regarding when data is expected to be on host already.

## A more detailed breakdown

1. `sequence_lengths.py` introduces the helper conversion function and type
2.  automodel and megatron`train.py` handle the conversion before passing to host-side loss and logprob code
3. `model_utils.py`, `algorithms/loss/utils.py` and `algorithms/loss/wrapper.py` explicitly accept CpuIntTuple and don't require .item() calls anymore. Also simplified some code that was recomputing seq_end as seq_start + padded_seq_lengths[seq_idx]
4. `models/megatron/data.py` normalizes inputs once at entry, and then use CPU code

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Comment by CarlosGomes98** [comment]
/ok to test [462d7e2](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/462d7e20f7660ddfdd467662e80e78e6aad315e9)

**Review by CarlosGomes98 — COMMENTED** [review]
From the /review-pr skill, as requested:

LGTM. Well-scoped refactor converting per-`.item()` D2H syncs into a single bulk `.tolist()` conversion at the host/device boundary via the new `to_cpu_int_tuple`/`CpuIntTuple` contract.

Verified:
- The wrapper simplification is exactly value-preserving: `logit_end = seq_end // cp_size` equals the old `(seq_start + padded_seq_lengths[seq_idx]) // cp_size` since `seq_end = padded_cu_seqlens[i+1]` and `padded_seq_lengths[i] = padded_cu_seqlens[i+1] - padded_cu_seqlens[i]`. Same holds for the fused-path `min(...)` branch. No off-by-one.
- Every production caller either converts via `to_cpu_int_tuple` at the boundary or normalizes internally at function entry; tuples are compatible with all downstream ops (`len()`, indexing, slicing, `//`).
- No circular-import risk (the new module imports only `torch` / `collections.abc`).
- The added `test_pack_input_ids_with_cpu_boundaries` is correct.

**Comment by CarlosGomes98** [comment]
/ok to test [ff3be49](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/ff3be4965975a8bf0b6a85c2d0c76b759de0bb24)

**Comment by CarlosGomes98** [comment]
/ok to test [6cb9b58](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/6cb9b5849f098dea82f921f675511853c8a182eb)

**Comment by CarlosGomes98** [comment]
/ok to test [b06fa80](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/b06fa8039a0847b3f86477f84835d395739049b8)

**Comment by CarlosGomes98** [comment]
/ok to test [58fb683](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/58fb6831c008895b03eac1cba15c59ef8daff479)

**Review by yuki-97 — COMMENTED** [review]
@bg51717 could you help review?

File: nemo_rl/algorithms/loss/utils.py [hunk 1]
```diff
@@ -345,8 +348,8 @@ def prepare_loss_input(
 
 def _pack_input_ids(
     input_ids: torch.Tensor,
-    cu_seqlens_q: torch.Tensor,
-    cu_seqlens_q_padded: torch.Tensor,
+    cu_seqlens_q: CpuIntTuple,
```
Comment (bg51717) [inline nemo_rl/algorithms/loss/utils.py:616]: This narrowing didn't reach `nemo_rl/models/megatron/train.py:326`, which still passes `packed_seq_params.cu_seqlens_q` / `.cu_seqlens_q_padded` as GPU tensors.

File: nemo_rl/algorithms/loss/utils.py [hunk 2]
```diff
@@ -118,7 +119,9 @@ def pack_rolled_draft_token_mask(
     per-segment left shift.
     """
     packed = _pack_input_ids(
```
Comment (bg51717) [inline nemo_rl/algorithms/loss/utils.py]: The boundaries passed here were already converted at `megatron/train.py:555/561`, so `_pack_input_ids` could take those existing tuples instead of re-deriving them.

File: tests/unit/algorithms/test_sequence_packing_fusion.py [hunk 3]
```diff
@@ -499,12 +500,27 @@ def test_pack_input_ids_in_bounds_lengths_unchanged():
...
+
+    input_ids = torch.arange(2 * 16).reshape(2, 16)
+    cu_q = (0, 11, 27)
+    cu_qp = (0, 16, 32)
+
+    actual = _pack_input_ids(
+        input_ids, cu_q, cu_qp, cp_rank=0, cp_size=1, roll_shift=-1
+    )
+    assert actual.shape == (1, 32)
+    assert torch.equal(actual[0, :10], input_ids[0, 1:11])
```
Comment (bg51717) [inline tests/unit/algorithms/test_sequence_packing_fusion.py]: This asserts 10 of 32 output positions, leaving sequence 0's padding tail and all of sequence 1 unchecked — the sibling tests at `:456-461` compare whole slots via `_rolled_padded_seq`.

File: nemo_rl/models/automodel/train.py [hunk 4]
```diff
@@ -648,11 +649,14 @@ def __call__(
         )
         # Wrap loss function for sequence packing if needed
         if self.enable_seq_packing:
+            cu_seqlens_q = to_cpu_int_tuple(
```
Comment (bg51717) [inline nemo_rl/models/automodel/train.py:652]: The packed logprob-unpacking loops at `:776` and `:986` still `.item()` per sequence off this same `flash_attn_kwargs.cu_seqlens_q`, as do `dtensor_policy_worker.py:1320` and `:1730`.

File: nemo_rl/models/megatron/train.py [hunk 5]
```diff
@@ -656,7 +666,7 @@ def __call__(
         self,
         data_dict: BatchedDataDict[Any],
         input_ids: torch.Tensor,
-        cu_seqlens_padded: torch.Tensor,
+        cu_seqlens_padded: Optional[torch.Tensor | CpuIntTuple],
```
Comment (bg51717) [inline nemo_rl/models/megatron/train.py:773]: `TopkLogitsPostProcessor` takes the same `cu_seqlens_padded` from the same dispatcher at `:376-382` but kept `torch.Tensor` and its per-sequence `int(...item())` loops, so the two sibling post-processors now disagree on the contract.

File: nemo_rl/models/megatron/train.py [hunk 6]
```diff
@@ -544,12 +545,21 @@ def __call__(
...
                 # cannot honor a custom prepare_fn (e.g. the value model's); guard
-                # rather than silently bypass it.
+                # before reading packed metadata so misconfiguration fails clearly.
                 assert self.prepare_fn is None, (
                     "sequence_packing.fuse_loss=true does not support a custom "
                     "prepare_fn (e.g. the value model's value-specific prep). "
                     "Disable fuse_loss for the value model."
                 )
+
+            cu_seqlens_q = to_cpu_int_tuple(packed_seq_params.cu_seqlens_q)
```
Comment (bg51717) [inline nemo_rl/models/megatron/train.py]: `cu_seqlens_q` / `cu_seqlens_q_padded` are only consumed by the `wrapper_cls(...)` call after the if/else, so the conversion could sit below it rather than splitting `if fuse_loss:` in two.

**Review by bg51717 — COMMENTED** [review]
LGTM.

Moves the sequence-packing boundary metadata (`cu_seqlens`, `seq_lengths`) to the host once, up front: `to_cpu_int_tuple()` converts before the model forward, and the loss/logprob code below it takes `CpuIntTuple` and indexes with plain Python ints instead of calling `.item()` per sequence.

Would be good to see a before/after sync count, given this is a `perf(...)` PR.

_Generated by Claude Code_

**Comment by CarlosGomes98** [comment]
/ok to test [84b45fe](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/84b45fe331704b291d7d4c10d1aa7ba1f7ee5e6e)

File: nemo_rl/algorithms/loss/utils.py (hunk 1 above)
Comment (CarlosGomes98) [reply nemo_rl/algorithms/loss/utils.py:616]: Added conversion at this point

File: tests/unit/algorithms/test_sequence_packing_fusion.py (hunk 3 above)
Comment (CarlosGomes98) [reply tests/unit/algorithms/test_sequence_packing_fusion.py]: added extra checks with _rolled_padded_seq

File: nemo_rl/models/automodel/train.py (hunk 4 above)
Comment (CarlosGomes98) [reply nemo_rl/models/automodel/train.py:652]: added those optimizations to automodel train.py as well

File: nemo_rl/models/megatron/train.py (hunk 5 above)
Comment (CarlosGomes98) [reply nemo_rl/models/megatron/train.py:773]: extended the functionality to topklogits

File: nemo_rl/models/megatron/train.py (hunk 6 above)
Comment (CarlosGomes98) [reply nemo_rl/models/megatron/train.py]: addressed

**Comment by CarlosGomes98** [comment]
/ok to test [350f904](https://github.com/NVIDIA-NeMo/RL/pull/3932/commits/350f9041afd765afe9584c727b45f17a95f59e16)

File: nemo_rl/models/automodel/train.py [hunk 7]
```diff
@@ -767,11 +771,14 @@ def __call__(
                 dtype=token_logprobs.dtype,
                 device=token_logprobs.device,
             )
-            cu_seqlens = processed_inputs.flash_attn_kwargs.cu_seqlens_q
+            cu_seqlens = to_cpu_int_tuple(
```
Comment (RayenTian) [inline nemo_rl/models/automodel/train.py:774]: Non-blocking: Could we consistently use `_cpu` for local variables created at the conversion boundary, as the Megatron [logprob](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/train.py#L794) and [top-k](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/train.py#L1055) paths already do?

  The remaining locations are:

  - `automodel/train.py`: `cu_seqlens_q` at [L652](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/automodel/train.py#L652), and `cu_seqlens` at [L774](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/automodel/train.py#L774) and [L986](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/automodel/train.py#L986).
  - `dtensor_policy_worker.py`: `cu_seqlens_q` at [L922](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/policy/workers/dtensor_policy_worker.py#L922), and `cu_seqlens` at [L1319](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/policy/workers/dtensor_policy_worker.py#L1319) and [L1725](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/policy/workers/dtensor_policy_worker.py#L1725).
  - `megatron/train.py`: `cu_seqlens_q` and `cu_seqlens_q_padded` at [L657–658](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/train.py#L657-L658).
  - `megatron/data.py`: `lengths_list` at [L1070](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/data.py#L1070), which could be named `seq_lengths_cpu`.

  This would make the converted CPU values easier to distinguish from the original tensors. Downstream parameters already annotated as `CpuIntTuple`, such as those in [the loss wrapper](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/algorithms/loss/wrapper.py#L54-L55), can keep their existing names.

File: nemo_rl/utils/sequence_lengths.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+# limitations under the License.
+
+from collections.abc import Sequence
+
+import torch
+
+CpuIntTuple = tuple[int, ...]
+
+
+def to_cpu_int_tuple(values: torch.Tensor | Sequence[int]) -> CpuIntTuple:
```
Comment (RayenTian) [inline nemo_rl/utils/sequence_lengths.py:22]: Would it make sense to apply the same bulk CPU conversion to these paths as well?

  - `TeacherFullPayloadPostProcessor`: the [CP gather loop](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/train.py#L964-L965) and [unpack loop](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/megatron/train.py#L999-L1000).
  - The value worker's `_unpack_value_sequences`: the [CP gather loop](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/value/workers/megatron_value_worker.py#L169-L170) and [unpack loop](https://github.com/NVIDIA-NeMo/RL/blob/350f9041afd765afe9584c727b45f17a95f59e16/nemo_rl/models/value/workers/megatron_value_worker.py#L180-L181).

  These still call `.item()` per sequence, so they seem like candidates for the same optimization. Would you prefer to include them here or leave them for a follow-up?

File: nemo_rl/utils/sequence_lengths.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    """Normalize sequence metadata at the host/device API boundary.
+
```
Comment (RayenTian) [inline nemo_rl/utils/sequence_lengths.py:24]: Optional, non-blocking: Would you consider grouping the CPU sequence metadata into one object and creating it during microbatch preparation? This could make ownership clearer and avoid passing separate boundary arguments through each layer.

For example, the structure could look like this:

```python
@dataclass(frozen=True)
class CpuSequenceMetadata:
    lengths: CpuIntTuple
    cu_seqlens: CpuIntTuple
    cu_seqlens_padded: CpuIntTuple

    @classmethod
    def from_lengths(cls, lengths) -> "CpuSequenceMetadata":
        ...

    @classmethod
    def from_packed_boundaries(
        cls, *, lengths, cu_seqlens, cu_seqlens_padded
    ) -> "CpuSequenceMetadata":
        ...
```

`from_lengths()` could handle packing without padding. `from_packed_boundaries()` could snapshot the final Megatron boundaries after all padding adjustments. The original sample lengths should remain separate from those adjusted boundaries.

The interfaces could then follow one pattern:

- **Automodel `ProcessedInputs`**: carry `cpu_metadata`, built in `process_microbatch()` and reused by training, logprob, and top-k paths.
- **Megatron `ProcessedInputs` / `ProcessedMicrobatch`**: carry the same `cpu_metadata` object alongside `packed_seq_params`. The separate `cu_seqlens_padded` field could be removed.
- **Legacy DTensor worker**: create the metadata during packing preparation and reuse it for loss, logprob, and top-k processing. The v2 worker would use the shared Automodel path.
- **Megatron postprocessors**: accept `cpu_metadata` and pass it through to loss wrappers.
- **Packing loss wrappers**: store the metadata object instead of separate CPU boundary fields.
- **Draft loss wrapper**: additionally accept `packed_seq_params` for GPU indexing.

For example:

```python
class SequencePackingLossWrapper:
    def __init__(self, loss_fn, prepare_fn, cpu_metadata, ...):
        self.cpu_metadata = cpu_metadata
        ...

class SequencePackingFusionLossWrapper:
    def __init__(self, loss_fn, prepare_fn, cpu_metadata, ...):
        self.cpu_metadata = cpu_metadata
        ...

class DraftLossWrapper:
    def __init__(
        self,
        ...,
        cpu_metadata=None,
        packed_seq_params=None,
    ):
        self.cpu_metadata = cpu_metadata
        self.packed_seq_params = packed_seq_params
        ...
```

Low-level packing and slicing helpers could still take individual tuples, extracted at the point of use. GPU boundaries needed by attention and draft indexing would remain in `packed_seq_params`.

This would keep CPU conversion in the preparation layer and give downstream consumers one consistent interface. Would this be worth considering, either here or in a follow-up?

**Comment by RayenTian** [comment]
Thanks for adding this! Overall LGTM. Left some comments.
